library/n2_protocol.hpp

```diff
@@ -13,36 +13,57 @@ struct data_in_file {
 	off_t local_offset = 0;
 	size_t fsize = 0;
 	int on_exit = 0;
+
+	void detach();
+	data_pointer read_and_detach();
 };
 
 struct data_place {
+	data_in_file in_file;
+	data_pointer in_memory;
+
 	enum place_type {
 		IN_MEMORY,
 		IN_FILE,
 	};
-
-	data_in_file in_file;
-	data_pointer in_memory;
-
 	place_type where() const;
 
-	static data_place from_file(const data_in_file &in_file);
-	static data_place from_memory(const data_pointer &in_memory);
+	void force_memory();
+
+	static data_place from_file(data_in_file in_file);
+	static data_place from_memory(data_pointer in_memory);
 };
 
 struct message {
```

Comment:
Why this structure now? Struct with virtual destructor and without methods looks strange.

Reply:
Without virtual destructor will be a bug, because `delete msg;` won't destruct derived correctly.

Reply:
Main question is `why this structure is needed?`

Reply:
@sabramkin explained me, that he use this struct
1. as alternative for `void*`. And he down casts pointer from `message*` to `read_request*` (as example).
2. as holder for `cmd`.
3. for virtual destructor and correct deletion.